src/source-updater.js

```diff
@@ -24,54 +24,144 @@ const actions = {
   },
   callback: (callback) => (type, sourceUpdater) => {
     callback();
+  },
+  duration: (duration) => (sourceUpdater) => {
+    sourceUpdater.mediaSource.duration = duration;
   }
 };
 
 const updating = (type, sourceUpdater) => {
   const sourceBuffer = sourceUpdater[`${type}Buffer`];
 
-  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queue[type].pending;
+  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queuePending[type];
 };
 
-const shiftQueue = (type, sourceUpdater) => {
-  const queue = sourceUpdater.queue[type];
+const processMediaSourceQueueEntry = (index, sourceUpdater) => {
+  const queueEntry = sourceUpdater.queue[index];
+  const types = queueEntry.type;
+  const bufferInUse =
+    types.reduce((acc, type) => acc || updating(type, sourceUpdater), false);
 
-  if (updating(type, sourceUpdater) || !queue.actions.length || !sourceUpdater.started_) {
-    return;
+  if (bufferInUse) {
+    // block until all buffers are ready for use
+    return false;
   }
 
-  const action = queue.actions.shift();
+  // remove the entry since it'll be processed
+  sourceUpdater.queue.splice(index, 1);
+  // when the entry needs multiple source buffer types, an individual type doesn't need to
+  // be passed for the action
+  queueEntry.action(sourceUpdater);
+
+  if (queueEntry.doneFn) {
+    queueEntry.doneFn();
+  }
 
-  queue.pending = action[1];
-  action[0](type, sourceUpdater);
+  return true;
 };
 
-const pushQueue = (type, sourceUpdater, action) => {
-  const queue = sourceUpdater.queue[type];
+const nextQueueIndexOfType = (type, queue) => {
+  for (let i = 0; i < queue.length; i++) {
+    const queueEntry = queue[i];
 
-  queue.actions.push(action);
-  shiftQueue(type, sourceUpdater);
+    if (Array.isArray(queueEntry.type)) {
```

Comment:
this is checking if the next entry in general is a media source entry, not if the next entry of this `type` is

Reply:
Yeah, since we have to block all types if there's a media source entry.

Reply:
might want to update the comment below then, since it's misleading